crates/core/tedge_api/src/entity_store.rs

```diff
@@ -676,6 +672,164 @@ impl EntityStore {
     }
 }
 
+/// In-memory representation of the entity tree
+struct EntityTree {
+    entities: HashMap<EntityTopicId, EntityNode>,
+    entity_id_index: HashMap<EntityExternalId, EntityTopicId>,
+}
+
+#[derive(Debug)]
+struct EntityNode {
+    metadata: EntityMetadata,
+    children: HashSet<EntityTopicId>,
+}
+
+impl EntityNode {
+    pub fn new(metadata: EntityMetadata) -> Self {
+        EntityNode {
+            metadata,
+            children: HashSet::new(),
+        }
+    }
+
+    pub fn metadata(&self) -> &EntityMetadata {
+        &self.metadata
+    }
+    pub fn mut_metadata(&mut self) -> &mut EntityMetadata {
+        &mut self.metadata
+    }
+}
+
+impl EntityTree {
+    /// Create the tree of entities for the main device, given its name, topic id and metadata
+    pub fn new(
+        device_name: EntityExternalId,
```

Comment:
Is this device name / external id required?

I kept it to reduce the amount of changes, but this external id is also given by the metadata.